### PR TITLE
feat: added structured errors

### PR DESCRIPTION
**Pull request by Ninjagod1251**
### What I did

Refactored the `except CompilerError as e: in compuile_project` a structured error response for failed tasks caused by `CompilerError`. This introduces a new Pydantic model to format the error and ensure it's compatible with frontend systems via FastAPI.

fixes: #50 

### How I did it

- Added a `CompilerErrorModel` class using Pydantic, which includes fields like `message`, `line`, `column`, and `errorType`.
- Updated the logic within the `compile_project` function to handle and format `CompilerError` in line with the new model.

### How to verify it

- Trigger a compilation task that raises a `CompilerError` by submitting a Vyper contract with a syntax error.
- Check the response from `/exceptions/{task_id}` to verify that the error details are returned in the correct format, as described in the issue.
- Ensure that the response structure matches the frontend expectations (JSON format with specific error fields).

### Checklist

- [] All changes are completed
- [] New test cases have been added for error handling and proper formatting of `CompilerError`.
- [ ] Documentation has been updated to reflect the changes in `/exceptions/{task_id}` endpoint behavior.

What still needs to check and verify

- Refactored the `/exceptions/{task_id}` endpoint to check specifically for tasks that failed due to a `CompilerError` and return a structured JSON response for the error.

### CODE REVIEW

**Comment by Ninjagod1251** [comment]
Im still using pdb to make sure the obj can parse the error. 

but it format good so far?

**Comment by Ninjagod1251** [comment]
and do we need to change the `get_task_exceptions` to handle the new format?

File: main.py [hunk 1]
```diff
@@ -12,8 +12,19 @@
...
 from pydantic import BaseModel
+from typing import Optional
 
 
+class CompilerErrorModel(BaseModel):
+    status: str = "failed"
+    message: str
+    column: Optional[int] = None
+    line: Optional[int] = None
+    error_type: str
```
Comment (fubuloubu) [inline main.py:23]: import this from `pydantic`
```suggestion
    error_type: str = Field(alias="errorType")
```

File: main.py [hunk 2]
```diff
@@ -12,8 +12,19 @@
...
+
+    class Config:
+        orm_mode = True
```
Comment (fubuloubu) [inline main.py:26]: I don't think is necessary
```suggestion
```

File: main.py [hunk 3]
```diff
@@ -192,6 +204,25 @@ async def compile_project(project_root: Path, manifest: PackageManifest):
         results[project_root.name] = compiled_manifest
         tasks[project_root.name] = TaskStatus.SUCCESS
     except CompilerError as e:
+        breakpoint()
+        # Convert the error details into the Pydantic model
+        error_details = [
+            CompilerErrorModel(
+                message=e.get('message', 'Unknown error'),
```
Comment (fubuloubu) [inline main.py:211]:
```suggestion
                message=e.get('message', str(e)),
```

File: main.py [hunk 4]
```diff
@@ -192,6 +204,25 @@ async def compile_project(project_root: Path, manifest: PackageManifest):
...
+                column=e.get('sourceLocation', {}).get('column', None),
+                line=e.get('sourceLocation', {}).get('line', None),
```
Comment (fubuloubu) [inline main.py:213]: by default, the 2nd arg to `.get()` on a `dict` is already `None`, so you don't need this
```suggestion
                column=e.get('sourceLocation', {}).get('column'),
                line=e.get('sourceLocation', {}).get('line'),
```

File: main.py [hunk 5]
```diff
@@ -192,6 +204,25 @@ async def compile_project(project_root: Path, manifest: PackageManifest):
...
+                error_type=e.__class__.__name__
+            ).dict()  # Converts to a dictionary for JSON serialization
```
Comment (fubuloubu) [inline main.py:215]: You don't need this if you change the FastAPI route that returns errors to be `-> list[CompilerErrorModel]`
```suggestion
            )
```

File: main.py [hunk 6]
```diff
@@ -138,13 +149,14 @@ async def get_task_status(task_id: str) -> TaskStatus:
     return tasks[task_id]
 
 
-@app.get("/exceptions/{task_id}")
+@app.get("/exceptions/{task_id}", response_model=list[CompilerErrorModel])
 async def get_task_exceptions(task_id: str) -> dict:
```
Comment (fubuloubu) [inline main.py:153]: If you use a typing annotation for return, FastAPI will automatically parse that into JSON using the model you give it (even inside a `list`)
```suggestion
@app.get("/exceptions/{task_id}")
async def get_task_exceptions(task_id: str) -> list[CompilerErrorModel]:
```

File: main.py [hunk 7]
```diff
@@ -12,8 +12,19 @@
...
 from fastapi.responses import HTMLResponse
 from pydantic import BaseModel
+from typing import Optional
 
 
+class CompilerErrorModel(BaseModel):
+    status: str = "failed"
+    message: str
+    column: Optional[int] = None
+    line: Optional[int] = None
```
Comment (fubuloubu) [inline main.py:22]: I think we are using 3.10 for this service and we don't need to use `Optional` anymore
```suggestion
    column: int | None = None
    line: int | None = None
```

File: main.py [hunk 8]
```diff
@@ -12,8 +12,19 @@
 from fastapi.openapi.docs import get_swagger_ui_html, get_swagger_ui_oauth2_redirect_html
 from fastapi.responses import HTMLResponse
 from pydantic import BaseModel
+from typing import Optional
 
 
+class CompilerErrorModel(BaseModel):
```
Comment (fubuloubu) [inline main.py:18]: oh, and more of a stylistic choice but I don't usually add `Model` after my models haha (although that conflicts with `ape.exceptions.CompilerError`)

think about what name would be most useful to see for downstream integrators to understand what this model contains

File: main.py [hunk 9]
```diff
@@ -192,6 +201,25 @@ async def compile_project(project_root: Path, manifest: PackageManifest):
...
+        results[project_root.name] = error_details
+        tasks[project_root.name] = TaskStatus.FAILED
+    
+    except CompilerError as e:
+        # convert into a pydantic model class subclass base model and every single field that i list will be a filed propery of model and conver to js via fastapi
+        # js schema convert into python checmo for pydancic 
+        breakpoint()
         results[project_root.name] = [
             f"{e['sourceLocation'].get('file', 'Unknown file')}\n{e['type']}: {e.get('formattedMessage', e['message'])}"
             for e in e.base_err.error_dict
```
Comment (fubuloubu) [inline main.py:225]: This path will never get triggered because it is handled above

**Review by fubuloubu — COMMENTED** [review]
Make sure to handle the line below containing `except Exception as e` using `CompilerErrorResponse`

File: main.py [hunk 10]
```diff
@@ -192,11 +200,27 @@ async def compile_project(project_root: Path, manifest: PackageManifest):
...
     except Exception as e:
-        results[project_root.name] = {e.__class__.__name__:str(e)}
+        # Handle any other exceptions that occur
+        generic_error_response = CompilerErrorResponse(
+            message=str(e),
+            column=None,  # Adjust if applicable
+            line=None,    # Adjust if applicable
+            error_type=e.__class__.__name__
+        )
+        results[project_root.name] = {e.__class__.__name__: str(e)}
```
Comment (fubuloubu) [inline main.py:225]: This should be a list

I wonder I'd this was the root cause of Joe's issue...

**Pull request merged**